### ISSUE TITLE
Dictionary: tapping issues which affects the Reader

### ISSUE THREAD

**Issue by gerhaher** [koreader/koreader#7473]
* KOReader version: v2021.03-12-gde543b3_2021-03-31/ (firmware 4.26.16704)
* Device: KA1

#### Issue
I made a film that shows what happen after the issue is triggered:
https://drive.google.com/file/d/1gPhUXikzMRKls7pmfyRK9gNZKMKrbUDz/view?usp=sharing
What you see is first that I can't change page with a single tap. Instead I need to double tap to do that. And when I tap << & >> I end up in first and last dictionary (and not in next).
Second, when I go back to the Reader, when I tap on the screen the dictionary opens again, every time. And when I swipe, well the dictionary is there again, intruding.
What is not shown in the video is that I can page backward/forward in the Reader by triple tapping.

So in the end I have to make a forced shut down of the device.

I noticed the issue with v2021.03-6-g1f15184_2021-03-30/, I didn't see it with v2021.03-3-ge334769_2021-03-29/ (although it could be because I didn't trigger it)

#### Steps to reproduce
I trigger the issue by using the dictionary. How many words you have to look up varies.
Before writing this bug report I did the following test:
I started every time from a powered off device.
And then I longtapped the same word ('ulcerous') from the same book over and over again until it was triggered.
Like this:
1. Long tap on 'ulcerous'
2. Tap once 'page forward' in the dictionary to check if the issue has been triggered
3. close dictionary window
Repeat 1... and so on

I did the test 4 times. The result was: the issue was triggered after 1 attempt, 13 attempts, 4 attempts, +50 attempts

verbose debug log: [crash1.log](https://github.com/koreader/koreader/files/6235266/crash1.log)

**Comment by NiLuJe** [comment]
Did you enable double-taps at one point, and if so, what's your double-tap interval set to?

**Comment by gerhaher** [comment]
> New question: what's your hold interval set to?

Default value (500)

**Comment by NiLuJe** [comment]
Does things behave better if you disable the BackgroundRunner plugin?

**Comment by gerhaher** [comment]
No, it happened again.

**Comment by NiLuJe** [comment]
Note to self: I'm *guessing* nothing clears the slot in whatever weird protocol the KA1 is using, which means the hold detection always passes.

Have to run for a bit, but I'll get you something to test later tonight (e.g., have the hold detection actually re-check the delta instead of relying on a previous clear).

**Comment by NiLuJe** [comment]
In the meantime, if you have shell access, I wouldn't mind having a look at a strace log of what happens:

`strace -fittvTy -e trace=%clock,_newselect,timerfd_create,timerfd_settime -p $(pidof reader.lua | sed -e 's/ / -p /g')`

**Comment by NiLuJe** [comment]
FWIW, things start to go awry at `03/31/21-10:40:21`, and I can't really find any reason why, so I'm *definitely* going to need a strace :/.

**Comment by NiLuJe** [comment]
Also, mildly curious why AutoSuspend trips `onInputEvent` *twice* (I smell ZMQs shenanigans).

Do you have KOSync enabled?

**Comment by gerhaher** [comment]
> Do you have KOSync enabled?

No

> I'm _definitely_ going to need a strace

And I'm _definitely_ need detailed instructions how to make one. :-)
Similar what you did a while ago about something else: https://github.com/koreader/koreader/issues/6582#issuecomment-683297132
(And I have wifi now)

**Comment by NiLuJe** [comment]
Well, thanks for digging that up, because it's basically exactly the same thing, except you stop at 7 and replace the command with the one I posted [above](https://github.com/koreader/koreader/issues/7473#issuecomment-811183642) ;o).

(Do try again on today's nightly, though, it might have fixed it).

**Comment by NiLuJe** [comment]
Barring that, sticking printfs everywhere helps, but a strace would be way more readable ;).

```diff
diff --git a/frontend/device/input.lua b/frontend/device/input.lua
index b0b12cf70..ac4544d1e 100644
--- a/frontend/device/input.lua
+++ b/frontend/device/input.lua
@@ -932,6 +932,7 @@ end
 -- it's either nil (meaning block forever waiting for input), or the earliest UIManager deadline (in most cases, that's the next scheduled task,
 -- in much less common cases, that's the earliest of UIManager.INPUT_TIMEOUT (currently, only KOSync ever sets it) or UIManager.ZMQ_TIMEOUT if there are pending ZMQs).
 function Input:waitEvent(now, deadline)
+    print("Input:waitEvent", now:tonumber(), deadline:tonumber())
     -- On the first iteration of the loop, we don't need to update now, we're following closely (a couple ms at most) behind UIManager.
     local ok, ev
     -- Wrapper around the platform-specific input.waitForEvent (which itself is generally poll-like).
@@ -957,17 +958,21 @@ function Input:waitEvent(now, deadline)
                     -- We use the ultimate deadline, as the kernel will just signal us when the timer expires during polling.
                     poll_deadline = deadline
                     with_timerfd = true
+                    print("timerfd, use input deadline", poll_deadline:tonumber())
                 else
                     if not deadline then
                         -- If we don't actually have a full timeout deadline, just honor the timer's.
                         poll_deadline = self.timer_callbacks[1].deadline
                         deadline_is_timer = true
+                        print("no input deadline, use timer deadline", poll_deadline:tonumber())
                     else
                         if self.timer_callbacks[1].deadline < deadline then
                             poll_deadline = self.timer_callbacks[1].deadline
                             deadline_is_timer = true
+                            print("timer < input; use timer deadline", poll_deadline:tonumber())
                         else
                             poll_deadline = deadline
+                            print("timer > input; use input deadline", poll_deadline:tonumber())
                         end
                     end
                 end
@@ -986,6 +991,7 @@ function Input:waitEvent(now, deadline)
                         poll_timeout = TimeVal:new{ sec = 0 }
                     end
                 end
+                print("w/ timer: effective deadline", poll_timeout:tonumber())
 
                 local timerfd
                 ok, ev, timerfd = input.waitForEvent(poll_timeout and poll_timeout.sec, poll_timeout and poll_timeout.usec)
@@ -994,22 +1000,26 @@ function Input:waitEvent(now, deadline)
 
                 -- If we've drained all pending input events, causing waitForEvent to time out, check our timers
                 if ok == false and ev == C.ETIME then
+                    print("ETIME")
                     -- Check whether the earliest timer to finalize a Gesture detection is up.
                     local consume_callback = false
                     if timerfd then
                         -- If we were woken up by a timerfd, that means the timerfd backend is in use, of course,
                         -- and it also means that we're guaranteed to have reached its deadline.
                         consume_callback = true
+                        print("timerfd, consume callback")
                     elseif not with_timerfd then
                         -- On systems where the timerfd backend is *NOT* in use, we have a few more cases to handle...
                         if deadline_is_timer then
                             -- We're only guaranteed to have blown the timer's deadline
                             -- when our actual select deadline *was* the timer's!
                             consume_callback = true
+                            print("deadline timer, consume callback")
                         elseif TimeVal:now() >= self.timer_callbacks[1].deadline then
                             -- But if it was a task deadline instead, we to have to check the timer's against the current time,
                             -- to double-check whether we blew it or not.
                             consume_callback = true
+                            print("deadline blown, consume callback")
                         end
                     end
 
@@ -1054,6 +1064,7 @@ function Input:waitEvent(now, deadline)
                     poll_timeout = TimeVal:new{ sec = 0 }
                 end
             end
+            print("w/o timer: effective deadline", poll_timeout:tonumber())
 
             ok, ev = input.waitForEvent(poll_timeout and poll_timeout.sec, poll_timeout and poll_timeout.usec)
         end -- if #timer_callbacks > 0
diff --git a/frontend/ui/uimanager.lua b/frontend/ui/uimanager.lua
index 9b4304ca9..a86b3d477 100644
--- a/frontend/ui/uimanager.lua
+++ b/frontend/ui/uimanager.lua
@@ -1588,22 +1588,27 @@ function UIManager:handleInput()
     local deadline
     -- Default to INPUT_TIMEOUT (which may be nil, i.e. block until an event happens).
     local wait_us = self.INPUT_TIMEOUT
+    print("UIManager: INPUT", wait_us)
 
     -- If we have any ZMQs registered, ZMQ_TIMEOUT is another upper bound.
     if #self._zeromqs > 0 then
         wait_us = math.min(wait_us or math.huge, self.ZMQ_TIMEOUT)
+        print("UIManager: ZMQ", wait_us)
     end
 
     -- We pass that on as an absolute deadline, not a relative wait time.
     if wait_us then
         deadline = now + TimeVal:new{ usec = wait_us }
+        print("UIManager: INPUT+ZMQ", deadline:tonumber())
     end
 
+    print("UIManager: wait_until, deadline", wait_until and wait_until:tonumber(), deadline and deadline:tonumber())
     -- If there's a scheduled task pending, that puts an upper bound on how long to wait.
     if wait_until and (not deadline or wait_until < deadline) then
         --             ^ We don't have a TIMEOUT induced deadline, making the choice easy.
         --                             ^ We have a task scheduled for *before* our TIMEOUT induced deadline.
         deadline = wait_until
+        print("UIManager: TASK", deadline:tonumber())
     end
 
     -- If allowed, entering standby (from which we can wake by input) must trigger in response to event
```

**Comment by gerhaher** [comment]
I've tried latest nightly and (so far) have failed to trigger it. So maybe it's fixed.

About the strace:
Isn't there a way to see all info (from start to finish) from a session in Putty.
When I started a session it went so fast and there was so much info and I didn't know how to stop it - and when it finally stopped after I shut down my device, I could see only the 2 last seconds. So this is what I got: 
[putty.txt](https://github.com/koreader/koreader/files/6245411/putty.txt)

**Comment by NiLuJe** [comment]
I don't have PuTTY on hand right now, but it probably defaults to a "tiny" scrollback buffer, so try expanding that to much larger (say, 15000 lines).

That said, that output does show erroneous behavior, so I wouldn't mind a look at the matching crash.log ;).

**Comment by NiLuJe** [comment]
Oh, as for stopping strace: <kbd>Ctrl</kbd>+<kbd>C</kbd>

**Comment by gerhaher** [comment]
I can't find that setting in Putty. Well, nevermind.

Matching crash.log: 
[crash.txt](https://github.com/koreader/koreader/files/6245603/crash.txt)

**Comment by NiLuJe** [comment]
Was that on today's nightly?

**Comment by gerhaher** [comment]
No

**Comment by NiLuJe** [comment]
Oh, okay. I'm mildly reassured ;).

Sidebar: that was written in the log, I'm an idiot ^^.
Also: it wasn't in verbose debug mode anyway ;).

Could you get me a strace on *today's* nightly, then, I'd like to double-check ;).

**Comment by NiLuJe** [comment]
As for PuTTy's scrollback setting: https://lmgtfy.app/?q=Putty+scrollback+buffer+setting ;).

**Comment by gerhaher** [comment]
> Could you get me a strace on today's nightly, then, I'd like to double-check ;).

[Put.txt](https://github.com/koreader/koreader/files/6245800/Put.txt)

**Comment by NiLuJe** [comment]
Okay, formatting aide, that looks saner, thanks :).

**Comment by NiLuJe** [comment]
And there was indeed an issue being the multiple onInputEvent handlers: [HookContainer was broken](https://github.com/NiLuJe/koreader/commit/f4152ee7f0929b33d8e4058eefc32b5f7c9d5439).

**Comment by NiLuJe** [comment]
Assuming this was fixed in #7478 & #7483 :).
